src/sentry/services/hybrid_cloud/rpc/endpoints.py

```diff
@@ -0,0 +1,236 @@
+            return False
+        if not getattr(v, "__isabstractmethod__"):
+            return False
+        return True
+
+
+@dataclasses.dataclass
+class MethodEndpoint:
+    params_serializer: JsonSerializer
+    result_serializer: JsonSerializer
+    method_name: str
+
+
+endpoint_registry: MutableMapping[str, ServiceEndpoint] = {}
+
+
+def expose_as_region_silo_rpc(
+    service_name: str, f: Callable[[], ServiceInterface]
+) -> Callable[[], ServiceInterface]:
+    if service_name in endpoint_registry:
+        raise ValueError(f"Service {service_name} registered twice!")
+
+    endpoint_registry[service_name] = ServiceEndpoint(
+        factory=f, service_name=service_name, silo_mode=SiloMode.REGION
+    )
+    return f
+
+
+def impl_with_region_silo_client(
+    service_name: str, service_base: Type[ServiceInterface]
+) -> Callable[[], ServiceInterface]:
+    pass
+
+
+class RpcClient(abc.ABC):
+    @abc.abstractmethod
+    def route(self, method_name: str, parameters: Mapping[str, Any]) -> BaseSiloClient:
+        pass
+
+    def invoke(
+        self, method_name: str, parameters: Mapping[str, Any], result_dataclass: Type[_T]
+    ) -> _T:
+        pass
+
+
+class RegionRpcClient(Generic[ServiceInterface], RpcClient):
+    service_name: str
+    service_base: ServiceInterface
+
+    def __init__(self, service_name: str, service_base: ServiceInterface):
+        self.service_name = service_name
+        self.service_base = service_base
+
+    def route(self, method_name: str, parameters: Mapping[str, Any]) -> BaseSiloClient:
+        if not hasattr(self.service_base, method_name):
+            raise NoMethodError(f"No method for rpc {self.service_name} {repr(method_name)}")
+
+        method = getattr(self.service_base, method_name)
+        org_routing: Callable[..., str | int] | None = getattr(method, "__org_routing")
+
+        org: str | int = org_routing(**parameters)
+        region_name: str | None
+        if isinstance(org, str):
+            region_name = OrganizationMapping.find_region_name_by_org_slug(org)
+        else:
+            region_name = OrganizationMapping.find_region_name_by_org_id(org)
+
+        if region_name is None:
+            raise MissingOrgError(
+                f"Organization {repr(org)} does not exist in OrganizationMapping."
+            )
+
+        try:
+            region = get_region_by_name(region_name)
+        except RegionResolutionError as e:
+            raise MissingRegionError(str(e))
+
+        return RegionSiloClient(region)
+
+
+class HasOrganizationId(Protocol):
+    organization_id: int
+
+
+class HasOrganizationSlug(Protocol):
+    organization_slug: str
+
+
+class HasRegionName(Protocol):
+    region_name: str
+
+
+_HasOrganizationId = TypeVar("_HasOrganizationId", bound=HasOrganizationId)
+_HasOrganizationSlug = TypeVar("_HasOrganizationSlug", bound=HasOrganizationSlug)
+_HasRegionName = TypeVar("_HasRegionName", bound=HasRegionName)
+
+
+def routes_region_by_org_id(
```

Comment:
Here's the trick.  For region silo endpoints WE NEED to be able ot determine the region to route rpc calls to.  This means either having callers explicitly figure this out, or somehow annotating service methods to converting paramters -> something that we can look up the region by.

This approach allows one to decorate service methods by matching the parameters of the callable against a protocol (say, routing by checking that the parameters as organization_id or organization_slug or region_name explicitly).  The problem is that this REQUIRES that the service method only take one method argument and that the method argument to enforce the type checking.

This might be solvable with python 3.10 which introduces better multi argument parameter type inference (`typing.ParamsSpec`)?  But unclear.

Reply:
I think i would 
a) put this comment in code
b) move these into the hybrid cloud service definitions by InterfaceWithLifecycle